src/transformers/models/paligemma/modeling_paligemma.py

```diff
@@ -306,24 +311,42 @@ def _merge_input_ids_with_image_features(self, image_features, inputs_embeds, in
             image_mask.unsqueeze(-1).expand_as(final_embedding), scaled_image_features
         )
         final_embedding = torch.where(pad_mask_expanded, torch.zeros_like(final_embedding), final_embedding)
+        if attention_mask is not None:
+            position_ids = (attention_mask.cumsum(-1)).masked_fill_((attention_mask == 0), 1)
+        else:
+            position_ids = None
 
-        final_attention_mask_4d = attention_mask.unsqueeze(1).unsqueeze(2) * attention_mask.unsqueeze(1).unsqueeze(-1)
-        final_attention_mask_4d = final_attention_mask_4d.float().expand(
-            -1, self.config.text_config.num_key_value_heads, -1, -1
-        )
-
-        # position_ids = torch.arange(0, sequence_length, device=input_ids.device).expand(batch_size, -1)
-        # position_ids = torch.where(input_ids == self.pad_token_id, torch.ones_like(position_ids), position_ids)
-        position_ids = (attention_mask.cumsum(-1)).masked_fill_((attention_mask == 0), 1)
+        if token_type_ids is not None:
+            # from llama modeling
+            target_length = cache_position[-1] + 1
+            causal_mask = torch.full(
+                (sequence_length, target_length), fill_value=min_dtype, dtype=dtype, device=device
+            )
+            if sequence_length != 1:
+                causal_mask = torch.triu(causal_mask, diagonal=1)
+            causal_mask *= torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)
+            causal_mask = causal_mask[None, None, :, :].expand(inputs_embeds.shape[0], 1, -1, -1)
+            if attention_mask is not None:
+                causal_mask = causal_mask.clone()  # copy to contiguous memory for in-place edit
+                mask_length = attention_mask.shape[-1]
+                padding_mask = causal_mask[:, :, :, :mask_length] + attention_mask[:, None, None, :]
+                causal_mask[:, :, :, :mask_length] = causal_mask[:, :, :, :mask_length].masked_fill(
+                    token_type_ids[:, None, None, :] == 0, 0
+                )
+                padding_mask = padding_mask == 0
+                causal_mask[:, :, :, :mask_length] = causal_mask[:, :, :, :mask_length].masked_fill(
+                    padding_mask, min_dtype
+                )
 
-        if labels is not None:
             final_labels = torch.full(
                 (batch_size, sequence_length), self.config.ignore_index, dtype=input_ids.dtype, device=input_ids.device
             )
             final_labels = torch.where(input_ids != self.pad_token_id, labels, final_labels)
```

Comment:
would it make sense to do
`final_labels = torch.where(token_type_ids == 1, labels, final_labels)`
here?

Reply:
hmm, now labels is passed frmo the processor as 
```python
labels = inputs["input_ids"].masked_fill(inputs["token_type_ids"] == 1, -100)
```
so this part is just used to mask out pad tokens it seems

Reply:
I see that now!  I was looking at old code. btw,
```labels = inputs["input_ids"].masked_fill(inputs["token_type_ids"] == 1, -100)```
->
```labels = inputs["input_ids"].masked_fill(inputs["token_type_ids"] == 0, -100)```

Reply:
(also btw pad_tokens appear to have value 0 in the returned ids, i feel that should be a different value but in any case for now it'd get masked by the line you mentioned. Maybe good to explicitly masked based on pad_token_id if that changes in the future tho)

Reply:
token types ids should not interfere with padding tokens . Padding <=> attention mask

Reply:
yes good point!